### PR TITLE
(#102) Create script to migrate data to S3 bucket.

### LINKED ISSUE #102: Enabler: Migrate data to S3
# Migrate existing print pages from SQL Server to S3 bucket.

Version 2 of CTS.Print removes the connection to the MS SQL Server database, however printable pages need to remain available for 90 days after being generated. In order to fill this gap, we need to create a script to retrieve the generated page and their metadata (trial IDs and search criteria) from the existing SQL Server database and store them in S3.

> **ESTIMATE** TBD

### PR DESCRIPTION
**Pull request by blairlearn**
Expects CSV file as exported by MS SQL Server Management Studio. (See https://github.com/NCIOCPL/clinical-trials-search-print/wiki/Deployment#data-migration)

Closes #102

### CODE REVIEW

File: migration/migrate.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import boto3
+from botocore.exceptions import ClientError
+
+S3_CLIENT = boto3.client("s3")
+
+if len(sys.argv) != 2:
+    raise RuntimeError("Please provide the input file name as a command line argument.")
+
+
+if (
```
Comment (jfrank-nih) [inline migration/migrate.py:34]: I think you can condense to:

`not os.environ.get("CTS_BUCKET_NAME", "").strip()`

With the caveat that a `None` os.environ object (I don't think that's possible?) will throw an error.

File: migration/migrate.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    or os.environ["CTS_BUCKET_NAME"].strip() == ""
+):
+    raise RuntimeError("The 'CTS_BUCKET_NAME' environment variable has not been set.")
+
+BUCKET = os.environ["CTS_BUCKET_NAME"]
+
+loadedCount = 0
+errorCount = 0
+totalCount = 0
+with open(sys.argv[1], encoding="utf-8-sig") as datafile:
```
Comment (jfrank-nih) [inline migration/migrate.py:46]: For clarity I would put `sys.argv[1]` into a variable `filename` up at line 32.

File: migration/migrate.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    csv.field_size_limit(sys.maxsize)
```
Comment (jfrank-nih) [inline migration/migrate.py:47]: Just curious--why is it needed?

File: migration/migrate.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    reader = csv.reader(datafile)
+
+    for row in reader:
+
+        key = row[0]
```
Comment (jfrank-nih) [inline migration/migrate.py:52]: Purely a "if you want to minimize your code" option. I believe this works:

```python
key, cacheDate, trialIDs, searchParams, content = row[0:4]
```

File: migration/migrate.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        cacheDate = row[1]
+        trialIDs = row[2]
+        searchParams = row[3]
+        content = row[4]
+
+        metadata = {}
```
Comment (jfrank-nih) [inline migration/migrate.py:58]: I would find this cleaner:

```python
metadata = {
    "migrated-data": "True",
    "originally-generated": cacheDate,
    "search-criteria": searchParams,
    "trial-id-list": trialIDs
}
```

File: migration/migrate.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+errorCount = 0
+totalCount = 0
+with open(sys.argv[1], encoding="utf-8-sig") as datafile:
+    csv.field_size_limit(sys.maxsize)
+    reader = csv.reader(datafile)
+
+    for row in reader:
+
+        key = row[0]
+        cacheDate = row[1]
```
Comment (jfrank-nih) [inline migration/migrate.py:53]: Typically python would do `cache_date` instead of `cacheDate`, per PEP8.

File: migration/migrate.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    or os.environ["CTS_BUCKET_NAME"] is None
+    or os.environ["CTS_BUCKET_NAME"].strip() == ""
+):
+    raise RuntimeError("The 'CTS_BUCKET_NAME' environment variable has not been set.")
+
+BUCKET = os.environ["CTS_BUCKET_NAME"]
+
+loadedCount = 0
+errorCount = 0
+totalCount = 0
```
Comment (jfrank-nih) [inline migration/migrate.py:45]: Technically `totalCount` isn't needed and could just be len(reader), right?

**Review by jfrank-nih — APPROVED** [review]
Linter has comments:

```
************* Module migrate
migrate.py:43:0: C0103: Constant name "loadedCount" doesn't conform to UPPER_CASE naming style (invalid-name)
migrate.py:44:0: C0103: Constant name "errorCount" doesn't conform to UPPER_CASE naming style (invalid-name)
migrate.py:45:0: C0103: Constant name "totalCount" doesn't conform to UPPER_CASE naming style (invalid-name)
migrate.py:85:15: W0718: Catching too general exception Exception (broad-exception-caught)
```

The constants aren't actually constants, so they could be ignored. Generally, if you were to use a default function `__main__` then this wouldn't pop.

I see the other is a non-AWS catch-all-problems, which is fair. 

So for both you could add an exclusion comment for `black` to make those go away.

File: migration/migrate.py (hunk 3 above)
Comment (jfrank-nih) [reply migration/migrate.py:47]: The `sys.maxsize` specifically.`

Comment (blairlearn) [reply migration/migrate.py:47]: The default is too small for some of the larger multi-trial pages.

**Pull request merged**